### ISSUE TITLE
Which mAP result should I pick?

### ISSUE THREAD

**Issue by shivagncs** [ultralytics/ultralytics#4527]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
I found a similar question but didn't get the answer to my question there, so I am asking a fresh one.

### Question

I am running a multiclass (5) object detection training using YOLOv8 and my questions is regarding which of the mAP results mentioned below I should pick. I am using yolov8m.pt and training for 80 epochs.

When training , I am passing in "iou=0.1" to the train command FYI, not sure if this is relevant but just mentioning

In the results.csv file, the "metrics/mAP50(B)" column lists epoch-wise values , and the highest value there is 0.66428(occurring during epoch 63)
<img width="101" alt="image" src="https://github.com/ultralytics/ultralytics/assets/138953776/af325360-dc7d-455b-a2c0-54f3bbf6cb2f">

However , in the "PR_curve.png" plot from the results folder, the "All classes mAP @0.5" values is listed as 0.659.
<img width="362" alt="image" src="https://github.com/ultralytics/ultralytics/assets/138953776/c1cb14f7-c6b9-410e-9327-66e3532065e2">

What is the difference between these, and which of these values is more appropriate to use as a result for mAP?

Appreciate your response!

**Comment by glenn-jocher** [comment]
Hello,

Great question! These two metrics are indeed very similar, but there is a subtle difference between them. The mAP (mean Average Precision) values listed in the "metrics/mAP50(B)" column of the results.csv file represent the values evaluated at the end of each training epoch. Therefore, maximum value observed (0.66428 at epoch 63 in your case) indicates the highest mAP reached during training.

On the other hand, the mAP value shown in the PR_curve.png plot ("All classes mAP @shivagncs.5" value: 0.659) is the mAP calculated on the validation data using the final model obtained after all training epochs have been completed.

Both values are valid measures of your model's performance, but they serve different purposes. The mAP value from the results.csv file is useful for tracking performance improvements over the course of training. Meanwhile, the mAP value from the PR_curve is beneficial for understanding the performance of the fully trained model on unseen data.

In terms of which one to pick, it depends on your objectives. If you are interested in the absolute best performance obtained at any point during training, the highest value from the results.csv file might be your choice. Conversely, if you are interested in the performance of your fully trained model on new, unseen data, you might prefer the value from the PR_curve.

I hope this clarifies your question! Please let us know if you have further questions.

**Comment by shivagncs** [comment]
Thank you very much for clarifying the difference. In my case, I am interested in the "best" version of the model as referenced by the "best.pt" file that gets written out after the model training run is completed. I will then be exporting this best.pt as an ONNX file and deploying it to an inference service for real world use. 
It seems that the "best.pt" file would correspond to the highest mAP value achieved during training(basically the value observed at epoch 63). Would you agree with this?

**Comment by glenn-jocher** [comment]
@shivagncs yes, your understanding is correct! The "best.pt" file generated by YOLOv8 during the training process indeed corresponds to the model state that achieved the highest mAP value on the validation set during any epoch throughout the training. Therefore, in your case, the "best.pt" file should represent the model state at the 63rd epoch where the maximum mAP of 0.66428 was recorded. 

This makes the "best.pt" file an excellent choice for deployment in your inference service, as it should theoretically provide the best detection performance among all the epochs, according to the validation set. Please keep in mind that this statement holds as long as the validation set is a good representation of the real-world data that the model will encounter during inference.

**Comment by shivagncs** [comment]
Perfect, thank you - Just one last question - Is the "last.pt" file better than the "best.pt" file in any way? I want to know where the last.pt file might be preferred to be used over best.pt since best.pt has the best mAP value

**Comment by glenn-jocher** [comment]
@shivagncs the "last.pt" file represents the state of your model at the very end of training, after the final epoch. This file can be useful for continuing training from the last point if needed, or for inspecting the final state of the model parameters.

The "best.pt" file, on the other hand, corresponds to the model state that achieved the highest mAP value on the validation set at any point during the entire training process.

In terms of model performance, the "best.pt" file is typically preferred as it represents the highest validation mAP achieved during training. However, there might be cases (although rare) where the "last.pt" might be preferred, for instance, if your model was still improving significantly at the end of training and might benefit from additional training time. But in the majority of scenarios, especially when the training curve has flattened or started to degrade due to overfitting, "best.pt" should provide the best performance based on your validation set.
